api/src/test/java/org/openmrs/module/idgen/integration/DuplicateIdentifiersPoolComponentTest.java

```diff
@@ -37,31 +38,30 @@ public void setUp() throws Exception {
     }
 
     @Test
-    @NotTransactional
-    public void testUnderLoad() throws Exception {
+    @Transactional(propagation = Propagation.NEVER)
+    public void testUnderLoad() {
 
         final List<String> generated = new ArrayList<String>();
 
-        List<Thread> threads = new ArrayList<Thread>();
+        List<Thread> threads = new ArrayList<>();
         for (int i = 0; i < NUM_THREADS; ++i) {
-            Thread thread = new Thread(new Runnable() {
-                @Override
-                public void run() {
-                    Context.openSession();
-                    Context.authenticate("admin", "test");
-                    IdentifierSource source = getService().getIdentifierSource(4);
-                    try {
-                        authenticate();
-                        sleep(100);
+            Thread thread = new Thread(() -> {
+                Context.openSession();
+                Context.authenticate("admin", "test");
```

Comment:
Can you explain why we needed to make all the changes here?

Reply:
Since multiple threads are writing to the same ArrayList at the same time, it’s not thread-safe. That’s why the test started failing randomly after the update. I fixed it by wrapping the add logic in a synchronized block.

Not totally sure why this didn’t happen before, though.